### PR TITLE
fix(ka): wire DetectedLabels from enrichment to DS catalog queries

### LINKED ISSUE #1052: KA tools do not pass DetectedLabels to DataStorage catalog queries
## Problem

The DataStorage API, ogen client, and scoring SQL all support `detected_labels` as a query parameter on `list_workflows` and `list_available_actions`. When provided, the scoring engine boosts GitOps-compatible workflows and penalizes non-GitOps ones via `buildDetectedLabelsBoostSQL` / `buildDetectedLabelsPenaltySQL`.

However, the KA tool implementations in `internal/kubernautagent/tools/custom/tools.go` (lines 192-198) only populate `ActionType`, `Severity`, `Component`, `Environment`, and `Priority` on the params struct. **`DetectedLabels` is never set**, meaning the GitOps-specific weighted scoring never activates through the catalog query path.

Today, enrichment labels only reach the LLM via prompt text — the LLM sees them but the catalog results are not pre-ranked by GitOps relevance.

## Impact

- Workflow catalog results are not scored/ranked by GitOps compatibility
- The full scoring infrastructure (DS-side) is unused
- LLM must rely solely on prompt context to prefer GitOps workflows, rather than receiving a pre-ranked list

## Evidence

- `tools.go` L192-198: only signal filters set, no `DetectedLabels`
- `oas_parameters_gen.go` L1779: `DetectedLabels OptString` exists on `ListWorkflowsByActionTypeParams`
- `scoring.go`: `buildDetectedLabelsBoostSQL` / `buildDetectedLabelsPenaltySQL` activate when `filters.DetectedLabels` is non-nil
- `workflow_discovery.go` L118: `WorkflowDiscoveryFilters.DetectedLabels` field exists

## Fix

Wire `DetectedLabels` from the enrichment result (available on `SignalContext` or via investigator context) into the `ListWorkflowsByActionTypeParams` before calling `ds.ListWorkflowsByActionType()` in both `list_workflows` and `list_available_actions` tools.

## Business Requirement

BR-AI-056 (GitOps-aware workflow selection)

### PR DESCRIPTION
**Pull request by jordigilh**
## Summary

Fixes #1052. Activates BR-AI-056 GitOps-aware workflow scoring by wiring `DetectedLabels` from enrichment through KA tools to DataStorage catalog queries.

- **Layer 1 (types.go):** Added `DetectedLabelsJSON` field to `SignalContext` as a pre-marshaled JSON transport
- **Layer 2 (investigator.go):** Marshal `enrichData.DetectedLabels` into `workflowSignal.DetectedLabelsJSON` before workflow selection, with error logging on marshal failure and debug logging on success (includes true label count and gitops_tool for SRE insight)
- **Layer 3 (tools.go):** Both `list_available_actions` and `list_workflows` tool Execute methods read `signal.DetectedLabelsJSON` and set `params.DetectedLabels` when non-empty

### Type consolidation (addresses Architecture review finding)

Unified three duplicate `DetectedLabels` struct definitions into a single canonical type in `pkg/shared/types/enrichment.go`. Both `enrichment.DetectedLabels` and `models.DetectedLabels` are now type aliases for `sharedtypes.DetectedLabels`, eliminating implicit JSON contract coupling. Removed 151 lines of duplicate code. `Scan`/`Value`/`IsEmpty`/`NewDetectedLabels` methods moved to the shared type (stdlib-only imports).

### Wire format change: `omitempty` on boolean fields

Added `omitempty` to all boolean JSON tags on the canonical `DetectedLabels` struct. This aligns with **DD-WORKFLOW-001 v1.5** which mandates: _"Booleans only included when true / False values are omitted from JSON."_ Previously `enrichment.DetectedLabels` emitted `"gitOpsManaged":false`; now false booleans are omitted. DS consumers handle this correctly:
- `discovery.go` only filters on `true` values
- `scoring.go` boosts on `true` values only; penalty SQL checks `IS NULL OR = 'false'` (handles absent correctly)

### Behavioral change: workflow ranking

After this fix, `list_workflows` and `list_available_actions` responses will return workflows ranked by GitOps compatibility (via DS scoring). Previously, results were ranked solely by `workflow_id ASC` tiebreaker since scoring was inactive. Specifically:

- GitOps-compatible workflows will be boosted to the top when the incident target is GitOps-managed
- Non-GitOps workflows may rank lower for GitOps-managed resources
- For non-GitOps resources (`gitOpsManaged: false`), scoring remains neutral (boost/penalty are 0.0)

## Test plan

- [x] **UT-KA-1052-001:** `list_available_actions` forwards `DetectedLabelsJSON` to DS params
- [x] **UT-KA-1052-002:** `list_workflows` forwards `DetectedLabelsJSON` to DS params
- [x] **UT-KA-1052-003:** `list_available_actions` omits `DetectedLabels` when empty
- [x] **UT-KA-1052-004:** `list_workflows` omits `DetectedLabels` when empty
- [x] **UT-KA-1052-005:** `DetectedLabelsJSON` round-trips through `SignalContext`
- [x] **UT-KA-1052-006:** nil enrichment `DetectedLabels` produces no param on tools
- [x] **IT-KA-1052-001:** Investigator wires enrichment `DetectedLabels` to DS query params (end-to-end Layer 1+2+3)
- [x] **IT-KA-1052-002:** Re-enrichment all-labels-failed preserves original labels on DS params
- [x] `go build ./...` clean
- [x] `go test -race` on all affected packages (0 failures, 0 races)
- [x] `go vet` on affected packages clean
- [x] `golangci-lint run --timeout=5m` — 0 new issues

### CODE REVIEW

**Comment by jordigilh** [comment]
## Multi-Persona Readiness Audit — PR #1054

**PR**: fix(ka): wire DetectedLabels from enrichment to DS catalog queries
**Issue**: #1052 / BR-AI-056
**Scope**: 3 production files (+25 lines), 2 new test files (+424 lines), 0 deletions

---

### 1. QE (Quality Engineering)

**Test Coverage Assessment:**
- 6 unit tests (UT-KA-1052-001 through 006) + 2 integration tests (IT-KA-1052-001, IT-KA-1052-002) — good breadth for a 3-file, 25-line change
- Happy path covered: labels forwarded to both `list_available_actions` and `list_workflows`
- Empty/nil edge cases covered: UT-003/004 (empty), UT-006 (nil/zero-value)
- Round-trip test (UT-005): context serialization verified
- Integration IT-002 covers the re-enrichment fallback path (RCA target not found → original labels preserved)

**Gaps / Concerns:**

- **MEDIUM — No malformed JSON test.** What happens if `json.Marshal` of `DetectedLabels` succeeds but produces JSON that DS rejects (e.g., unknown enum in `gitOpsTool`)? The investigator error branch (line 297) logs but continues silently — scoring is just "inactive." Acceptable for v1, but a test proving the error branch fires and the investigation still completes would add confidence.
- **LOW — No concurrency test.** `DetectedLabelsJSON` is set on a struct (not a shared resource), so not a real risk, but noted for completeness.
- **LOW — `allTools[0]` / `allTools[1]` positional indexing.** If tool registration order ever changes, tests silently test the wrong tool. Consider asserting `allTools[0].Name() == "list_available_actions"` as a guard.

**Verdict: PASS** (medium gap is low-risk — marshal failure is a defensive branch that degrades gracefully)

---

### 2. Production Security

**Data Flow Analysis:**
- `DetectedLabels` originates from K8s API discovery (annotations, labels) → marshaled to JSON → transported as a string field on `SignalContext` → forwarded as HTTP query parameter to DS

**Concerns:**

- **LOW — Query parameter injection.** JSON string passed as `OptString` via ogen-generated client (handles URL-encoding). DS uses `json.Unmarshal` into typed struct (`models.DetectedLabels`) with field validation. No raw SQL concatenation. Safe.
- **LOW — No input size limit.** `DetectedLabels` is a fixed-schema struct (~200-300 bytes marshaled). No unbounded fields.
- **NONE — No secrets in transit.** Contains only boolean flags and tool names (`"argocd"`, `"flux"`). No PII, no credentials.
- **NONE — No new API surface exposed.** `detected_labels` query parameter already existed in DS OpenAPI spec; this PR just populates it.

**Verdict: PASS**

---

### 3. Product / Business

**BR-AI-056 Alignment:**
- The issue correctly identifies that DS scoring infrastructure (boost/penalty SQL) existed but was never activated because KA tools didn't forward `DetectedLabels`. This PR closes that gap.
- Behavioral change clearly documented: GitOps-managed resources now get pre-ranked workflow results.
- Scoring is neutral for non-GitOps targets (boost=0, penalty=0), so non-GitOps paths are unaffected.

**Verdict: PASS**

---

### 4. Architecture

**Design Assessment:**
- **Three-layer pipeline is clean:** types.go (transport) → investigator.go (marshal) → tools.go (forward). Single responsibility per layer.
- **Pre-marshaled JSON string** (`DetectedLabelsJSON`) avoids re-marshaling per tool call and sidesteps import cycles between KA types and enrichment types.

**Concerns:**

- **MEDIUM — Schema coupling via implicit JSON contract.** `enrichment.DetectedLabels` is marshaled on KA side; `models.DetectedLabels` is unmarshaled on DS side. Two different Go types in different packages. If field names or JSON tags diverge, scoring silently breaks. Today they're compatible, but there's no compile-time or test-time contract enforcing this.
  - **Recommendation:** A cross-package round-trip test (marshal enrichment → unmarshal models) would catch drift. Could be a follow-up issue.
- **LOW — `string` transport vs typed transport.** Using `string` for `DetectedLabelsJSON` means the compiler can't catch type misuse. The alternative (importing enrichment types) would create import cycles, so string approach is pragmatic.

**Verdict: PASS** (schema coupling is the most actionable item — recommend a follow-up issue)

---

### 5. SRE / Observability

**Logging Assessment:**
- **Debug log (V=1)** on success: `"detected labels attached for workflow discovery scoring"` with `correlation_id` and `detected_labels_json_len` — good for tracing without leaking full label content at normal verbosity.
- **Error log** on marshal failure: includes `correlation_id` and the error itself. Sufficient for diagnosis.

**Concerns:**

- **LOW — No metric.** No counter tracking how often scoring activates vs. stays inactive. Fine for first pass; a `kubernaut_ka_detected_labels_forwarded_total` counter could be added later if scoring effectiveness needs monitoring.
- **NONE — Graceful degradation.** Marshal failure → scoring inactive → investigation proceeds normally. No crash, no retry loop.

**Verdict: PASS**

---

### 6. UX / Developer Experience

- Entirely backend — no user-facing UI change.
- Downstream effect: workflow ordering may change for GitOps-managed targets. Operators see more relevant workflows ranked higher. Intended UX improvement.

**Verdict: PASS**

---

### 7. API Compatibility

- No new API endpoints.
- `detected_labels` query parameter already existed in DS OpenAPI spec.
- `DetectedLabelsJSON` field added with `json:"detected_labels_json,omitempty"` — zero-value omitted, backward compatible.

**Verdict: PASS**

---

### Summary

| Persona | Verdict | High | Medium | Low |
|---|---|---|---|---|
| QE | PASS | 0 | 1 | 2 |
| Prod Security | PASS | 0 | 0 | 2 |
| Product | PASS | 0 | 0 | 0 |
| Architecture | PASS | 0 | 1 | 1 |
| SRE | PASS | 0 | 0 | 1 |
| UX | PASS | 0 | 0 | 0 |
| API Compat | PASS | 0 | 0 | 0 |

**Overall: PASS — 0 HIGH, 2 MEDIUM, 6 LOW**

### Recommended Follow-ups (not blocking merge):

1. **Cross-package schema contract test** — round-trip `enrichment.DetectedLabels` → marshal → unmarshal into `models.DetectedLabels` and assert field parity. Catches silent drift.
2. **Marshal failure test** — prove the error branch in investigator fires and the investigation completes without labels.
3. **Positional tool guard** — assert tool name before indexing `allTools[N]` in unit tests.

**Comment by jordigilh** [comment]
## Deep-Dive Readiness Audit (Second Pass) — PR #1054

The initial audit reviewed only the 5 files listed in the PR commit. A full `git diff origin/main...origin/fix/1052-wire-detected-labels-to-ds` reveals **8 files changed** (+513 / -163), including significant type unification work not described in the PR summary. This audit covers the full scope.

### Undocumented Changes (not in PR description)

The PR description mentions 3 layers (types.go, investigator.go, tools.go). The actual diff also includes:

1. **`pkg/shared/types/enrichment.go`** (+59 lines): Added `omitempty` to all boolean JSON tags on `DetectedLabels`. Moved `NewDetectedLabels()`, `IsEmpty()`, `Scan()`, and `Value()` methods here from `models/workflow_labels.go`. Added `database/sql/driver` import.
2. **`pkg/datastorage/models/workflow_labels.go`** (-151 lines): `DetectedLabels` replaced with type alias `= sharedtypes.DetectedLabels`. Removed duplicated struct, methods, and constructor.
3. **`internal/kubernautagent/enrichment/enricher.go`** (+3 / -14 lines): `DetectedLabels` replaced with type alias `= sharedtypes.DetectedLabels`. Removed duplicate struct definition.

**These are type unification changes that consolidate 3 separate `DetectedLabels` definitions into 1 canonical type in `pkg/shared/types`. This is a good architectural improvement but carries risk that should be explicitly called out.**

---

### 1. QE (Quality Engineering)

**Previous assessment confirmed** — 6 UT + 2 IT cover the wiring well.

**New concern:**

- **MEDIUM — No tests cover the `omitempty` JSON tag change.** The old `enrichment.DetectedLabels` emitted `"gitOpsManaged":false` (no omitempty); the new shared type emits nothing for `false`. While this is safe for DS (scoring only activates for `true` values, filtering only adds conditions for `true`), there are no tests verifying that a `DetectedLabels{GitOpsManaged: false}` still round-trips correctly through DS parsing. The existing scoring tests may cover this implicitly, but it's worth verifying.
- **LOW — `allTools[0]` / `allTools[1]` positional indexing** (reiterated from first pass).

**Verdict: PASS**

---

### 2. Production Security

**No change from first pass. PASS.**

---

### 3. Product / Business

**No change from first pass. PASS.**

---

### 4. Architecture

**Upgraded assessment — type unification is significant:**

- **POSITIVE — Single source of truth.** Three separate `DetectedLabels` structs (enrichment, models, shared) collapsed to one. Eliminates the implicit JSON contract coupling concern from the first audit — now KA and DS marshal/unmarshal the exact same Go type.
- **POSITIVE — Type alias approach.** `type DetectedLabels = sharedtypes.DetectedLabels` preserves method sets and avoids breaking existing callers.

**Concerns:**

- **MEDIUM — `omitempty` on boolean fields changes wire format.** Before: `{"gitOpsManaged":false,"pdbProtected":false,...}`. After: `{}` (all false booleans omitted). This is safe for current DS consumers (verified: `discovery.go` lines 340-347 only filter on `true`; `scoring.go` lines 91-166 only boost on `true`; penalty SQL line 196 checks `IS NULL OR = 'false'`). However, any future DS consumer that distinguishes "field absent" from "field is false" would break. The PR description should document this wire format change.
- **MEDIUM — `database/sql/driver` import in `pkg/shared/types`.** This package is imported by many consumers (KA, gateway, etc.) that don't need SQL scanning. The `Scan`/`Value` methods are only needed by DS. Consider moving them to a `pkg/shared/types/sql` subpackage or keeping them in `models/` via receiver-free helper functions. Not blocking, but increases coupling.
- **LOW — PR description underrepresents scope.** "+449 / -0" in the PR metadata vs actual "+513 / -163". The type unification deserves its own bullet in the summary.

**Verdict: PASS** (wire format change is safe given current consumers, but should be documented)

---

### 5. SRE / Observability

**No change from first pass. PASS.**

---

### 6. API Compatibility

- **Wire format change**: `DetectedLabels` JSON output for `false` booleans changed from explicit `false` to absent. DS parsing treats absent as `false` (Go zero value), so backward compatible.
- `Scan`/`Value` methods on `DetectedLabels` still work via type alias — PostgreSQL JSONB column reading/writing unaffected.

**Verdict: PASS**

---

### 7. Spec Compliance (DD-WORKFLOW-001)

- DD-WORKFLOW-001 v1.5 states: "Booleans only included when true / False values are omitted from JSON." The `omitempty` change actually **aligns** with the spec — the old code was non-compliant by emitting `false` values. This is a correctness improvement.

**Verdict: PASS — the change improves spec compliance.**

---

### Updated Summary

| Persona | Verdict | High | Medium | Low |
|---|---|---|---|---|
| QE | PASS | 0 | 1 (omitempty not tested) | 1 |
| Prod Security | PASS | 0 | 0 | 2 |
| Product | PASS | 0 | 0 | 0 |
| Architecture | PASS | 0 | 2 (wire format + sql/driver coupling) | 1 |
| SRE | PASS | 0 | 0 | 1 |
| API Compat | PASS | 0 | 0 | 0 |
| Spec Compliance | PASS | 0 | 0 | 0 |

**Overall: PASS — 0 HIGH, 3 MEDIUM, 5 LOW**

### Key Findings vs First Audit

1. **Type unification is a net positive** — eliminates the cross-package schema coupling concern from the first audit. Now KA and DS use the identical Go type.
2. **`omitempty` change aligns with DD-WORKFLOW-001 v1.5** — the old behavior (emitting `false` booleans) was technically non-compliant.
3. **PR description should be updated** to mention the type unification and wire format change.

### Recommended Follow-ups (not blocking merge)

1. ~~Cross-package schema contract test~~ — **no longer needed** (type alias eliminates the risk)
2. **Update PR description** to document the type unification and `omitempty` wire format change
3. **Consider moving `Scan`/`Value`** out of `pkg/shared/types` to avoid coupling non-DB consumers to `database/sql/driver`

File: internal/kubernautagent/investigator/investigator.go [hunk 1]
```diff
@@ -284,6 +284,21 @@ func (inv *Investigator) Investigate(ctx context.Context, signal katypes.SignalC
...
+	// so workflow discovery tools forward them to DS catalog queries, activating
+	// GitOps-aware scoring.
+	if enrichData != nil && enrichData.DetectedLabels != nil {
+		if dlJSON, err := json.Marshal(enrichData.DetectedLabels); err == nil {
+			workflowSignal.DetectedLabelsJSON = string(dlJSON)
+			inv.logger.V(1).Info("detected labels attached for workflow discovery scoring",
+				"correlation_id", correlationID,
+				"detected_labels_json_len", len(dlJSON))
+		} else {
+			inv.logger.Error(err, "failed to marshal detected labels for workflow discovery, scoring will be inactive",
```
Comment (jordigilh) [inline internal/kubernautagent/investigator/investigator.go:301]: **[QE — MEDIUM] No test covers this error branch.**

`json.Marshal` on a well-typed `DetectedLabels` struct will essentially never fail (no channels, funcs, or cycles), so this branch is defensive. However, a unit test that forces a marshal failure (e.g., via an interface mock that returns an unmarshalable type) and asserts the investigation still completes would close the gap.

Non-blocking — the graceful degradation (scoring inactive, investigation continues) is correct.

File: pkg/shared/types/enrichment.go [hunk 2]
```diff
@@ -148,7 +154,7 @@ type DetectedLabels struct {
 	// ========================================
```
Comment (jordigilh) [inline pkg/shared/types/enrichment.go:152]: **[Architecture — MEDIUM] `omitempty` added to boolean JSON tags changes wire format.**

Before: `{"gitOpsManaged":false,"pdbProtected":false,...}` (all booleans emitted).
After: `{}` (false booleans omitted).

This actually **aligns with DD-WORKFLOW-001 v1.5** which states: _"Booleans only included when true / False values are omitted from JSON."_ So this is a correctness fix.

Verified safe for current DS consumers:
- `discovery.go:340-347` — only filters on `true` values
- `scoring.go:91-166` — only boosts on `true` values
- `scoring.go:196` — penalty SQL checks `IS NULL OR = 'false'` (handles absent correctly)

Consider documenting this wire format change in the PR description since it affects all serialized `DetectedLabels` payloads.

File: pkg/shared/types/enrichment.go [hunk 3]
```diff
@@ -182,7 +188,46 @@ type DetectedLabels struct {
...
+		!d.PDBProtected &&
+		!d.HPAEnabled &&
+		!d.Stateful &&
+		!d.HelmManaged &&
+		!d.NetworkIsolated &&
+		d.ServiceMesh == ""
+}
+
+// Scan implements sql.Scanner for PostgreSQL JSONB column scanning.
+func (d *DetectedLabels) Scan(value interface{}) error {
```
Comment (jordigilh) [inline pkg/shared/types/enrichment.go]: **[Architecture — MEDIUM] `database/sql/driver` import couples non-DB consumers to SQL scanning.**

`pkg/shared/types` is imported by KA, gateway, and other packages that never need `Scan`/`Value`. The `database/sql/driver` import is stdlib so it's lightweight, but it conceptually couples a shared types package to database concerns.

Non-blocking suggestion: In a follow-up, consider moving `Scan`/`Value` to a `pkg/shared/types/sqlutil` subpackage or keeping them as standalone functions in `models/`.

File: test/unit/kubernautagent/tools/custom/detected_labels_1052_test.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+			ctx := katypes.WithSignalContext(contextBackground(), katypes.SignalContext{
+				Severity:           "critical",
+				ResourceKind:       "Deployment",
+				Environment:        "production",
+				Priority:           "P0",
+				DetectedLabelsJSON: detectedLabelsJSON,
+			})
+
+			allTools := custom.NewAllTools(fake)
+			listActions := allTools[0]
```
Comment (jordigilh) [inline test/unit/kubernautagent/tools/custom/detected_labels_1052_test.go:84]: **[QE — LOW] Positional tool indexing is fragile.**

`allTools[0]` assumes `list_available_actions` is always first. If `NewAllTools` registration order changes, this test silently tests the wrong tool.

Suggestion: add a guard assertion:
```go
Expect(allTools[0].Name()).To(Equal("list_available_actions"))
```

File: internal/kubernautagent/enrichment/enricher.go [hunk 5]
```diff
@@ -40,19 +41,9 @@ type OwnerChainEntry struct {
...
-	GitOpsTool               string   `json:"gitOpsTool"`
-	PDBProtected             bool     `json:"pdbProtected"`
-	HPAEnabled               bool     `json:"hpaEnabled"`
-	Stateful                 bool     `json:"stateful"`
-	HelmManaged              bool     `json:"helmManaged"`
-	NetworkIsolated          bool     `json:"networkIsolated"`
-	ServiceMesh              string   `json:"serviceMesh"`
-	ResourceQuotaConstrained bool     `json:"resourceQuotaConstrained"`
-}
+// DetectedLabels is the canonical label detection result type.
```
Comment (jordigilh) [inline internal/kubernautagent/enrichment/enricher.go:44]: **[Architecture — POSITIVE] Type unification via alias.**

Collapsing the duplicate `DetectedLabels` struct into a type alias for `sharedtypes.DetectedLabels` eliminates the implicit JSON contract coupling between KA and DS. Now both sides marshal/unmarshal the exact same Go type. Good cleanup.

File: pkg/datastorage/models/workflow_labels.go [hunk 6]
```diff
@@ -101,107 +101,14 @@ func (c *CustomLabels) Scan(value interface{}) error {
...
-
-	// ========================================
-	// RESOURCE CONSTRAINTS (#366, DD-HAPI-018 v1.4)
-	// ========================================
-
-	// ResourceQuotaConstrained indicates if any ResourceQuota exists in namespace
-	// Detection: List ResourceQuotas in namespace
-	// API Call: kubectl get resourcequota -n <namespace> -o json
-	ResourceQuotaConstrained bool `json:"resourceQuotaConstrained,omitempty"`
+// DetectedLabels is the canonical label detection result type (DD-WORKFLOW-001 v2.3).
```
Comment (jordigilh) [inline pkg/datastorage/models/workflow_labels.go]: **[Architecture — POSITIVE] Type alias consolidation.**

Replacing the 100+ line duplicate struct with `type DetectedLabels = sharedtypes.DetectedLabels` is a clean single-source-of-truth pattern. The `Scan`/`Value` methods defined on the shared type are automatically available through the alias.

Note: The PR description should mention this consolidation — it removes 151 lines and changes the canonical type location, which is architecturally significant.

File: internal/kubernautagent/investigator/investigator.go [hunk 7]
```diff
@@ -284,6 +284,21 @@ func (inv *Investigator) Investigate(ctx context.Context, signal katypes.SignalC
 			deletedResourceWarning(signalKind, signalName, signalNS))
 	}
 
+	// #1052 / BR-AI-056: Marshal enrichment DetectedLabels into the signal context
+	// so workflow discovery tools forward them to DS catalog queries, activating
+	// GitOps-aware scoring.
+	if enrichData != nil && enrichData.DetectedLabels != nil {
+		if dlJSON, err := json.Marshal(enrichData.DetectedLabels); err == nil {
+			workflowSignal.DetectedLabelsJSON = string(dlJSON)
```
Comment (jordigilh) [inline internal/kubernautagent/investigator/investigator.go:292]: **[SRE — LOW] Consider logging label count, not just JSON length.**

`detected_labels_json_len` gives byte size but not semantic insight. Adding a field count (e.g., how many booleans are `true`) would help SREs quickly assess whether scoring will be meaningful for a given incident without bumping log verbosity. Non-blocking.

File: internal/kubernautagent/tools/custom/tools.go [hunk 8]
```diff
@@ -146,6 +146,9 @@ func (t *listActionsTool) Execute(ctx context.Context, args json.RawMessage) (st
 		Environment: signal.Environment,
 		Priority:    ogenclient.ListAvailableActionsPriority(signal.Priority),
 	}
+	if signal.DetectedLabelsJSON != "" {
```
Comment (jordigilh) [inline internal/kubernautagent/tools/custom/tools.go:149]: **[Prod Security — LOW] Input is safe — just confirming.**

`signal.DetectedLabelsJSON` was marshaled by `json.Marshal` in the investigator (trusted source, well-typed struct). It's forwarded as `OptString` via the ogen client which handles URL-encoding. DS parses it with `json.Unmarshal` into a typed struct with validation. No injection vector.

**Review by jordigilh — COMMENTED** [review]
Inline comments for each audit finding. See the full audit in previous PR comments.

File: internal/kubernautagent/investigator/investigator.go (hunk 1 above)
Comment (jordigilh) [reply internal/kubernautagent/investigator/investigator.go:301]: Acknowledged. json.Marshal on a well-typed struct with no channels/funcs/cycles won't fail in practice, so this branch is purely defensive. The graceful degradation (scoring inactive, investigation continues) is the correct behavior.

If we ever want to test this, we'd need to inject a custom json.Marshaler interface, which would mean changing the production code structure just for testability of an unreachable path. Leaving as-is for now.

File: pkg/shared/types/enrichment.go (hunk 2 above)
Comment (jordigilh) [reply pkg/shared/types/enrichment.go:152]: Good catch on the DD-WORKFLOW-001 v1.5 alignment. Updated the PR description to document the wire format change explicitly, including the DS consumer safety analysis you outlined (discovery filters on true only, scoring penalty handles absent correctly).

File: pkg/shared/types/enrichment.go (hunk 3 above)
Comment (jordigilh) [reply pkg/shared/types/enrichment.go]: Agree with the concern. The stdlib `database/sql/driver` import is lightweight (no transitive deps), but conceptually it couples a shared types package to DB concerns. Tracked for follow-up -- moving `Scan`/`Value` to a `pkg/shared/types/dbutil` subpackage or back to standalone functions in models would be cleaner.

File: pkg/datastorage/models/workflow_labels.go (hunk 6 above)
Comment (jordigilh) [reply pkg/datastorage/models/workflow_labels.go]: Thanks. Updated the PR description to call out the consolidation explicitly: 3 duplicate structs -> 1 canonical type + 2 aliases, -151 lines.

File: internal/kubernautagent/investigator/investigator.go (hunk 7 above)
Comment (jordigilh) [reply internal/kubernautagent/investigator/investigator.go:292]: Addressed in b343803. Replaced `detected_labels_json_len` with `true_label_count` (count of boolean labels that are `true`) and `gitops_tool` (the detected tool name). This gives SREs immediate semantic insight into whether scoring will be meaningful without bumping verbosity.

File: internal/kubernautagent/tools/custom/tools.go (hunk 8 above)
Comment (jordigilh) [reply internal/kubernautagent/tools/custom/tools.go:149]: Confirmed. The trust chain is: typed Go struct -> json.Marshal (investigator) -> string field on SignalContext (in-process) -> OptString via ogen client (URL-encoded) -> json.Unmarshal into typed struct (DS handler). No external input touches this path.

**Comment by jordigilh** [comment]
## Re-Review: Comment Resolution Assessment

All 8 inline comments have been addressed. Here's the disposition:

### MEDIUM findings

| # | Finding | Resolution | Status |
|---|---|---|---|
| 1 | **[QE]** Error branch untested (`investigator.go:297`) | Team acknowledged as defensive/unreachable. Would require changing prod code for testability of an unreachable path. Reasoning is sound. | **Accepted — Won't Fix** |
| 2 | **[Arch]** `omitempty` wire format change (`enrichment.go:154`) | PR description updated with explicit wire format change section including DS consumer safety analysis. | **Resolved** |
| 3 | **[Arch]** `sql/driver` import coupling (`enrichment.go:214`) | Team acknowledged. Tracked for follow-up (move to `dbutil` subpackage). Stdlib-only, no transitive dep risk. | **Accepted — Follow-up** |

### LOW findings

| # | Finding | Resolution | Status |
|---|---|---|---|
| 4 | **[QE]** Positional tool indexing (`test:84`) | **Not addressed.** No guard assertion added. Still using `allTools[0]`/`allTools[1]` without name verification. | **Open — LOW risk** |
| 5 | **[SRE]** Log label count (`investigator.go:292`) | **Fixed in `b343803`.** Replaced `detected_labels_json_len` with `true_label_count` (variadic `countTrueLabels` helper) and `gitops_tool`. Clean implementation. | **Resolved** |
| 6 | **[Security]** Input safety (`tools.go:149`) | Confirmed by team. Trust chain validated. | **Resolved** |

### POSITIVE findings (no action needed)

| # | Finding | Status |
|---|---|---|
| 7 | Type alias unification (`enricher.go:44`) | Acknowledged |
| 8 | Type alias consolidation (`workflow_labels.go:104`) | PR description updated with consolidation details |

### New commit review: `b343803` (SRE logging fix)

The `countTrueLabels` helper is clean — variadic bool, simple loop. The log line now emits `true_label_count` and `gitops_tool`, which gives SREs immediate insight into scoring activation without parsing JSON. No concerns.

### Updated Summary

| Category | Before | After |
|---|---|---|
| MEDIUM open | 3 | 0 (2 resolved, 1 accepted) |
| LOW open | 2 | 1 (positional indexing) |
| Commits | 1 | 3 |

**Overall: PASS — Ready to merge.** The one remaining LOW (positional indexing) is non-blocking and does not affect correctness.
